drivers/input/input_cap12xx.c

```diff
@@ -0,0 +1,232 @@
+
+	cap12xx_process(data->dev);
+}
+
+static void cap12xx_isr_handler(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct cap12xx_data *data = CONTAINER_OF(cb, struct cap12xx_data, int_gpio_cb);
+
+	k_work_submit(&data->work);
+}
+
+#ifdef CONFIG_INPUT_CAP12XX_POLL
+static void cap12xx_timer_handler(struct k_timer *timer)
+{
+	struct cap12xx_data *data = CONTAINER_OF(timer, struct cap12xx_data, timer);
+
+	k_work_submit(&data->work);
+}
+#endif
+
+static int cap12xx_init(const struct device *dev)
+{
+	const struct cap12xx_config *config = dev->config;
+	struct cap12xx_data *data = dev->data;
+	int r;
+
+	if (!device_is_ready(config->i2c.bus)) {
+		LOG_ERR("I2C controller device not ready");
+		return -ENODEV;
+	}
+
+	data->dev = dev;
+
+	k_work_init(&data->work, cap12xx_work_handler);
+
+	if (config->int_gpio.port != NULL) {
+		if (!gpio_is_ready_dt(&config->int_gpio)) {
+			LOG_ERR("Interrupt GPIO controller device not ready");
+			return -ENODEV;
+		}
+
+		r = gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT);
+		if (r < 0) {
+			LOG_ERR("Could not confighure interrupt GPIO pin");
+			return r;
+		}
+
+		r = gpio_pin_interrupt_configure_dt(&config->int_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+		if (r < 0) {
+			LOG_ERR("Could not configure interrupt GPIO interrupt");
+			return r;
+		}
+
+		gpio_init_callback(&data->int_gpio_cb, cap12xx_isr_handler,
+				   BIT(config->int_gpio.pin));
+
+		r = gpio_add_callback(config->int_gpio.port, &data->int_gpio_cb);
+		if (r < 0) {
+			LOG_ERR("Could not set gpio callback");
+			return r;
+		}
+
+		r = cap12xx_clear_interrupt(&config->i2c);
+		if (r < 0) {
+			LOG_ERR("Could not clear interrupt");
+			return r;
+		}
+
+		r = cap12xx_enable_interrupt(&config->i2c, true);
+		if (r < 0) {
+			LOG_ERR("Could not configure interrupt");
+			return r;
+		}
+	}
+#ifdef CONFIG_INPUT_CAP12XX_POLL
+	else {
+		k_timer_init(&data->timer, cap12xx_timer_handler, NULL);
+
+		r = cap12xx_enable_interrupt(&config->i2c, false);
+		if (r < 0) {
+			LOG_ERR("Could not configure interrupt");
+			return r;
+		}
+
+		k_timer_start(&data->timer, K_MSEC(CONFIG_INPUT_CAP12XX_PERIOD),
+			      K_MSEC(CONFIG_INPUT_CAP12XX_PERIOD));
+	}
+#endif
+	if (config->repeat == 0) {
+		r = i2c_reg_write_byte_dt(&config->i2c, REG_REPEAT_ENABLE, REPEAT_DISABLE);
+		if (r < 0) {
+			LOG_ERR("Could not disable repeated events");
+			return r;
+		}
+		LOG_DBG("disabled repeated events");
+	} else {
+		r = i2c_reg_write_byte_dt(&config->i2c, REG_REPEAT_ENABLE, REPEAT_ENABLE);
+		if (r < 0) {
+			LOG_ERR("Could not enable repeated events");
+			return r;
+		}
+		LOG_DBG("enabled repeated events");
+	}
+
+	return 0;
+}
+
+#define CAP12XX_INIT(index)                                                                        \
+	BUILD_ASSERT(DT_INST_PROP_LEN(index, input_codes) == DT_INST_PROP(index, channels));       \
+	static const uint8_t cap12xx_input_channels = DT_INST_PROP(index, channels);               \
```

Comment:
you probably want to use `cap12xx_input_channels_##inst` as variable name or this would break if you had multiple instances